File: ledger-models-javascript/node/wrappers/models/security/IndexSecurity.ts
import Security from './security';
import { SecurityProto } from '../../../fintekkers/models/security/security_pb';
import { IndexTypeProto } from '../../../fintekkers/models/security/index/index_type_pb';

/**
 * Wrapper for Security messages whose product_type is a descendant of
 * INDEX in hierarchy.json (CPI_SERIES, SOFR_SERIES, EQUITY_INDEX, etc.).
 * The index-specific fields live in the index_details sub-message (one of
 * the non_bond_details oneof variants).
 */
class IndexSecurity extends Security {
  constructor(proto: SecurityProto) {
    super(proto);
  }

  /** Which index family this security represents. Defaults to
   * UNKNOWN_INDEX_TYPE when index_details is not populated. */
  getIndexType(): IndexTypeProto {
    const details = this.proto.getIndexDetails();
    return details ? details.getIndexType() : IndexTypeProto.UNKNOWN_INDEX_TYPE;
  }
}

export default IndexSecurity;

File: ledger-models-python/fintekkers/wrappers/models/security/index_security.py
"""IndexSecurity wrapper.

Specialises Security for index products (CPI-U, SOFR index, S&P 500, ...).
Index details live in the SecurityProto.non_bond_details oneof under
index_details.
"""

from fintekkers.models.security.index.index_type_pb2 import IndexTypeProto
from fintekkers.wrappers.models.security.security import Security


class IndexSecurity(Security):
    """Wraps a SecurityProto whose non_bond_details oneof carries
    index_details."""

    def get_index_type(self) -> IndexTypeProto:
        """Returns the IndexTypeProto value carried by this security's
        index_details. Returns UNKNOWN_INDEX_TYPE when no index_details is
        populated (e.g. the security is mis-typed as an index).
        """
        self._ensure_hydrated()
        if self.proto.WhichOneof("non_bond_details") != "index_details":
            return IndexTypeProto.UNKNOWN_INDEX_TYPE
        return self.proto.index_details.index_type
